rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+`nix-instantiate`) should not be changed to avoid breakage.
+
+Most `nix` subcommands work on a list of arguments called
+"installables" for lack of a better word. For example,
+
+    # nix run nixpkgs:hello dwarffs:dwarffs
+
+takes two flake-based installables. The general form is:
+
+    <flake-ref>(:<attr-path>)?
+
+Examples of installables:
+
+* `nixpkgs:packages.hello`
+* `nixpkgs:hello` - short for `nixpkgs:packages.hello`
+* `nixpkgs/release-19.03:hello` - overrides the Git branch to use
+* `github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082:hello` -
+  specifies the exact Git revision to use
+* `dwarffs` - short for `dwarffs:defaultPackage`
+* `nix:hydraJobs.build.x86_64-darwin`
+* `.:hydraJobs.build.x86_64-darwin` - refers to the flake in the
+  current directory (which can be a dirty Git tree)
+* `.` - short for `.:defaultPackage`
+
+If no argument is given, the default is `.`; thus,
+
+    # nix build
+
+is equivalent to
+
+    # nix build .:defaultPackage
+
+For backwards compatibility, it's possible to use non-flake Nix
+expressions using `-f`, e.g. `nix build -f foo.nix foo.bar`.
+
+## Lock files and reproducible evaluation
+
+Inputs specified in `flake.nix` are typically "unlocked": they are
+flake references that don't specify an exact revision (e.g. `nixpkgs`
+rather than
+`github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082`). To
+ensure reproducibility, Nix will automatically generate and use a
+*lock file* called `flake.lock` in the flake's directory. The lock
+file contains a tree of mappings from the flake references specified
+in `flake.nix` to direct flake references that contain revisions.
+
+For example, if `flake.nix` contains
+```
+inputs =
+  [ "nixpkgs"
+    github:edolstra/import-cargo
+  ];
+```
+then the resulting lock file might be:
+```
+{
+    "version": 2,
+    "inputs": {
+        "github:edolstra/import-cargo": {
+            "id": "import-cargo",
+            "inputs": {},
+            "narHash": "sha256-mxwKMDFOrhjrBQhIWwwm8mmEugyx/oVlvBH1CKxchlw=",
+            "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"
+        },
+        "nixpkgs": {
+            "id": "nixpkgs",
+            "inputs": {},
+            "narHash": "sha256-p7UqhvhwS5MZfqUbLbFm+nfG/SMJrgpNXxWpRMFif8c=",
+            "uri": "github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082"
+        }
+    }
+}
+```
+
+Thus, when we build this flake, the input `nixpkgs` is mapped to
+`github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e`. Nix
+will also check that the content hash of the input is equal to the one
+recorded in the lock file. This check is superfluous for Git
+repositories (since the commit hash serves a similar purpose), but for
+GitHub archives, we cannot directly check that the contents match the
+commit hash.
+
+Note that lock files are only used at top-level: the `flake.lock` files
```

Comment:
Yes, good point. We could simply always use the lock file of a dependency if it exist (so good practice would be not to commit lock files for library flakes). It gets tricky for flakes that provide both libraries and applications (like Nixpkgs)...

Reply:
This seems like something to consider carefully or it could become a big pain.

Reply:
I'm not sure I understand the difference between `packages` and `apps`, but what about having whether a lock file is considered be based on which set the reference is in?

Reply:
I think also that when a lock file is present, it should be used. It basically says "we provide a curated set, don't break it". It's then up to the community to ensure they don't use it incorrectly. 

Now, in practice I think that is not something we can always rely on and so we want to be able to override/patch. Also, because upstreams tend to pin things to much; just look at the amount of patching we need with Python applications. We are given the attribute`inputs`, which is already imported, and won't allow us to override or patch the Flake when needed which, if correct, is a limitation.

Reply:
@ryantm see e.g. the explanation given for Cargo, https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries

Reply:
Oh, right, this would be based on the inputs and `packages`/`apps` are outputs.

Reply:
Lockfiles shouldn't be committed to repositories under any circumstance, unless it is a distribution repository. This is no different to [vendor bundling](https://fedoraproject.org/wiki/Bundled_Libraries), which is rightly opposed. It's for a distributor to decide how to combine dependencies.

I agree that lockfiles in dependencies should be ignored. If this principle were abandoned, it should be strictly opt-in.

Reply:
Could someone explain why lock files in dependencies should be ignored? Doesn't it break the reproducibility of evaluation if the inputs of the dependency got updated?

Reply:
Without care, there is a tension between flakes offering composability on the one hand and reproducibility on the other. Ignoring dependent lockfiles is one way of reducing that tension: The flake.nix should specify the *interface* for the flake (eventually with version bounds or similar?) and the flake.lock should specify the *current composition* for people working on that project directly.

For example, I'd like us to eventually have a `stdenv` flake that abstracts out the bootstrapping logic and definition of mkDerivation from nixpkgs proper. When I'm developing on `stdenv` itself, I want some *fixed* definitions of things like gcc etc. so I have something to work with, but the entire point is that I ultimately want the *caller* to specify these! So when nixpkgs imports the stdenv flake, it should provide the pieces that make it all work.

Reply:
Maybe it might make sense, for a first draft, to say “lock files in dependencies are a hard error for now” and to just add that as a to-be-resolved-later question?

I'm not sure we need what @FRidh wants right now (and it's something that apparently only makes sense for applications, so not that well-explored for build systems at least), and this way is forwards-compatible with any future road we might end up choosing for handling lock files in dependencies.